### PR TITLE
[WebGL] Implement packed ScatterND

### PR DESCRIPTION
**Pull request by Linchenn**
After this PR, 'USE-batch30' model would gain 30%~50% performance improvement. The 'Packed-ScatterND' column shows this PR's performance on 'USE-batch30'.

image

Specifically, on MacBook Pro, the time that 'USE-batch30' spends in 'ScatterNd' drops from 62.84 ms to 15.01ms.

image
image

To see the logs from the Cloud Build CI, please join either our [discussion](https://groups.google.com/a/tensorflow.org/forum/#!forum/tfjs) or [announcement](https://groups.google.com/a/tensorflow.org/forum/#!forum/tfjs-announce) mailing list.

- - -
This change is [Reviewable](https://reviewable.io/reviews/tensorflow/tfjs/7292)

### CODE REVIEW

File: tfjs-backend-webgl/src/scatter_packed_gpu.ts [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+              }
+              if (flattenedIndex[0] == coords[0] + 1) {
+                sum.zw += updVals.xy;
+                found.zw = vec2(1.);
+              }
+              if (flattenedIndex[1] == coords[0] + 1) {
+                sum.zw += updVals.zw;
+                found.zw = vec2(1.);
+              }
+            }
```
Comment (Linchenn) [inline tfjs-backend-webgl/src/scatter_packed_gpu.ts:95]: The 4 if-branches seem to hurt performance, so I tried to replace it with the vectorized codes as the following and pasted tests:
```js
vec4 isMatched = 1. - vec4(bvec4(flattenedIndex[0] - coords[0],
                                 flattenedIndex[1] - coords[0],
                                 flattenedIndex[0] - coords[0] - 1,
                                 flattenedIndex[1] - coords[0] - 1));
if (dot(isMatched, vec4(1.)) > 0.) {
  vec4 updVals = ${updatesSnippet};
  found += isMatched.xxzz + isMatched.yyww;
  sum += updVals.xyxy * isMatched.xxzz + updVals.zwzw * isMatched.yyww;
}
...
setOutput(mix(${defaultValueSnippet}, sum, vec4(bvec4(found))));
```

However, it does not show obvious improvements (the performace is recorded in 'Packed-ScatterND-vectorizedBranches' column in PR description), so we could use the 4-if-branches here, which has better readability.

Comment (mattsoulanille) [reply tfjs-backend-webgl/src/scatter_packed_gpu.ts:95]: The four if statements is certainly more readable. I'm a bit surprised it doesn't hurt performance.

**Review by mattsoulanille — APPROVED** [review]
LGTM

**Review by pyu10055 — CHANGES_REQUESTED** [review]
Thank you! is this feature enabled by default and fully tested with current tests?

> *[Reviewable](https://reviewable.io/reviews/tensorflow/tfjs/7292#-:-NMG4RFK2agbbNM43RHB:b-khbsc5)* status: :shipit: complete! 1 of 1 approvals obtained (waiting on @Linchenn)

___
*[`tfjs-backend-webgl/src/scatter_packed_gpu.ts` line 3 at r1](https://reviewable.io/reviews/tensorflow/tfjs/7292#-NMFkpeC4holRaw4yiWt:-NMFkpeC4holRaw4yiWu:bb6ht38) ([raw file](https://github.com/tensorflow/tfjs/blob/15e3e9417b70a743d2cd410431814390f28d62f7/tfjs-backend-webgl/src/scatter_packed_gpu.ts#L3)):*
> ```XML
> /**
>  * @license
>  * Copyright 2018 Google LLC. All Rights Reserved.
> ```

2023

File: tfjs-backend-webgl/src/scatter_packed_gpu.ts (hunk 1 above)
Comment (pyu10055) [reply tfjs-backend-webgl/src/scatter_packed_gpu.ts:95]: is the performance indifference happening on mobile as well?

**Review by pyu10055 — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/tensorflow/tfjs/7292)* status: :shipit: complete! 1 of 1 approvals obtained (waiting on @Linchenn and @mattsoulanille)

___
*[`tfjs-backend-webgl/src/scatter_packed_gpu.ts` line 76 at r7](https://reviewable.io/reviews/tensorflow/tfjs/7292#-NMG8GhnAYuQVfKuEPFN:-NMG8Gho3jEUqI0oTZL-:b-qf268m) ([raw file](https://github.com/tensorflow/tfjs/blob/910a1503f8cd085cdd48ce3e874666aa1dae6ce1/tfjs-backend-webgl/src/scatter_packed_gpu.ts#L76)):*
> ```XML
>               }
>             }
>             if (flattenedIndex[0] == coords[0] || flattenedIndex[1] == coords[0] ||
> ```

is this check necessary? given the contained the branches will check again?

File: tfjs-backend-webgl/src/scatter_packed_gpu.ts (hunk 1 above)
Comment (Linchenn) [reply tfjs-backend-webgl/src/scatter_packed_gpu.ts:95]: Yes, the chart in PR description shows our mobile devices.

**Review by Linchenn — COMMENTED** [review]
I have tested it locally, for both current changes and the vectorized branch optimization through:
```
yarn test --test_verbose_timeout_warnings --verbose_failures --nocache_test_results  --//:grep='scatterND'
```
and tested 'USE-30' model correctness on the local benchmark tool.

I am running a nightly test now https://pantheon.corp.google.com/cloud-build/builds/f79d985c-2a59-4d2c-98c6-cc904231334a?project=learnjs-174218. Will not until it is passed.

To be more safe, we could add a 'WEBGL_PACK_SCATTERND' (default as false) at first?

> *[Reviewable](https://reviewable.io/reviews/tensorflow/tfjs/7292#-:-NMGo9Q03lNaO28O-YWH:b-qcka8k)* status: :shipit: complete! 1 of 1 approvals obtained (waiting on @mattsoulanille and @pyu10055)

___
*[`tfjs-backend-webgl/src/scatter_packed_gpu.ts` line 3 at r1](https://reviewable.io/reviews/tensorflow/tfjs/7292#-NMFkpeC4holRaw4yiWt:-NMGnufb5sOLsqkzKvyp:bizduzk) ([raw file](https://github.com/tensorflow/tfjs/blob/15e3e9417b70a743d2cd410431814390f28d62f7/tfjs-backend-webgl/src/scatter_packed_gpu.ts#L3)):*
<details><summary><i>Previously, pyu10055 (Ping Yu) wrote…</i></summary><blockquote>

2023
</blockquote></details>

I have updated it. It's weird that reviewable tools sometimes does not show the latest changes.
___
*[`tfjs-backend-webgl/src/scatter_packed_gpu.ts` line 76 at r7](https://reviewable.io/reviews/tensorflow/tfjs/7292#-NMG8GhnAYuQVfKuEPFN:-NMGnNrp3ClIJKten4xN:b-fbarh) ([raw file](https://github.com/tensorflow/tfjs/blob/910a1503f8cd085cdd48ce3e874666aa1dae6ce1/tfjs-backend-webgl/src/scatter_packed_gpu.ts#L76)):*
<details><summary><i>Previously, pyu10055 (Ping Yu) wrote…</i></summary><blockquote>

is this check necessary? given the contained the branches will check again?
</blockquote></details>

Yes, the immediate following line `vec4 updVals = ${updatesSnippet};` is a read instruction. Only if this check is passed, the read instruction would be executed. This has visible, even though small ~2ms for the model, improvements.



_Code quote:_
```XML
vec4 updVals = ${updatesSnippet};
```

**Review by pyu10055 — APPROVED** [review]
> Reviewed 2 of 3 files at r4, 1 of 1 files at r5, 1 of 1 files at r8, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/tensorflow/tfjs/7292)* status: :shipit: complete! 2 of 1 approvals obtained (waiting on @mattsoulanille)

**Review by pyu10055 — COMMENTED** [review]
LGTM, given it is guarded under WEBGL_PACK flag.

> *[Reviewable](https://reviewable.io/reviews/tensorflow/tfjs/7292#-:-NMVW28s2EZiFG98vVgH:b-ru234a)* status: :shipit: complete! 2 of 1 approvals obtained (waiting on @mattsoulanille)

**Review by pyu10055 — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/tensorflow/tfjs/7292)* status: :shipit: complete! 1 of 1 approvals obtained (waiting on @Linchenn and @mattsoulanille)

___
*[`tfjs-backend-webgl/src/scatter_packed_gpu.ts` line 76 at r7](https://reviewable.io/reviews/tensorflow/tfjs/7292#-NMG8GhnAYuQVfKuEPFN:-NMVgOAzCZ2QXutl4r7t:bkqi424) ([raw file](https://github.com/tensorflow/tfjs/blob/910a1503f8cd085cdd48ce3e874666aa1dae6ce1/tfjs-backend-webgl/src/scatter_packed_gpu.ts#L76)):*
<details><summary><i>Previously, Linchenn wrote…</i></summary><blockquote>

Yes, the immediate following line `vec4 updVals = ${updatesSnippet};` is a read instruction. Only if this check is passed, the read instruction would be executed. This has visible, even though small ~2ms for the model, improvements.
</blockquote></details>

can you move this to each child branches to avoid this top level branch statement?

**Review by Linchenn — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/tensorflow/tfjs/7292)* status: :shipit: complete! 1 of 1 approvals obtained (waiting on @mattsoulanille and @pyu10055)

___
*[`tfjs-backend-webgl/src/scatter_packed_gpu.ts` line 76 at r7](https://reviewable.io/reviews/tensorflow/tfjs/7292#-NMG8GhnAYuQVfKuEPFN:-NMVpspE1ELMHH2qLAwD:bm8cvov) ([raw file](https://github.com/tensorflow/tfjs/blob/910a1503f8cd085cdd48ce3e874666aa1dae6ce1/tfjs-backend-webgl/src/scatter_packed_gpu.ts#L76)):*
<details><summary><i>Previously, pyu10055 (Ping Yu) wrote…</i></summary><blockquote>

can you move this to each child branches to avoid this top level branch statement?
</blockquote></details>

Merged mutual exclusive branches as if...else... . Thank you Ping for thinking deeply about it!

**Review by pyu10055 — APPROVED** [review]

**Review by mattsoulanille — APPROVED** [review]

**Comment by Linchenn** [comment]
Just completed [nightly tests](https://pantheon.corp.google.com/cloud-build/builds/ddc47514-34bc-4b02-8347-f18b779a62f9;step=38?project=learnjs-174218).

**Pull request merged**